### PR TITLE
jit: isolate and document per-build tag specific definitions 

### PR DESCRIPTION
**Pull request by mathetake**
This is the continuation of refactor on jit pkg. Notably, 
this commit isolates per-build tag definition into arch_*
pkg and provide build-tag free documentation interface
on `arch.go`.

### CODE REVIEW

File: internal/wasm/jit/arch.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	// newArchContext returns a new archContext which is architecture-specific type to be embedded in callEngine.
+	newArchContext func() archContext
+
+	// jitcall is used by callEngine.execWasmFunction and the entrypoint to enter the JITed native code.
+	// codeSegment is the pointer to the initial instruction of the compiled native code.
+	// ce is "*callEngine" as uintptr.
+	jitcall func(codeSegment, ce uintptr)
+
+	// newCompiler returns a new compiler interface which can be used to compile the given function instance.
+	// The function returned must be invoked when finished compiling, so use `defer` to ensure this.
```
Comment (codefromthecrypt) [inline internal/wasm/jit/arch.go:21]: "The function returned must be invoked when finished compiling," doesn't seem to make sense.. isn't this a compilation function (so invoking it is compiling?)

File: internal/wasm/jit/arch_amd64.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+
+	asm "github.com/twitchyliquid64/golang-asm"
+	"github.com/twitchyliquid64/golang-asm/obj/x86"
+
+	wasm "github.com/tetratelabs/wazero/internal/wasm"
+	"github.com/tetratelabs/wazero/internal/wazeroir"
+)
+
+func init() {
```
Comment (codefromthecrypt) [inline internal/wasm/jit/arch_amd64.go:13]: helps clarify when looking at files as not always notice the file name.

```suggestion
// init initializes variables for the amd64 architecture
func init() {
```

File: internal/wasm/jit/arch.go (hunk 1 above)
Comment (mathetake) [reply internal/wasm/jit/arch.go:21]: oh thanks this is the remnants after moving the lock to engine.go.. will fix

File: internal/wasm/jit/arch_arm64.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"math"
+
+	asm "github.com/twitchyliquid64/golang-asm"
+	"github.com/twitchyliquid64/golang-asm/obj/arm64"
+
+	wasm "github.com/tetratelabs/wazero/internal/wasm"
+	"github.com/tetratelabs/wazero/internal/wazeroir"
+)
+
+func init() {
```
Comment (codefromthecrypt) [inline internal/wasm/jit/arch_arm64.go:14]: helps clarify when looking at files as not always notice the file name.

```suggestion
// init initializes variables for the arm64 architecture
func init() {
```

File: internal/wasm/jit/arch_arm64_test.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package jit
+
+import (
+	"testing"
+	"unsafe"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestArchContextOffsetInEngine(t *testing.T) {
```
Comment (codefromthecrypt) [inline internal/wasm/jit/arch_arm64_test.go:10]:
```suggestion
func TestArchContextOffsetInArm64Engine(t *testing.T) {
```

File: internal/wasm/jit/arch_other.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+//go:build !amd64 && !arm64
+
+package jit
+
+type archContext struct{}
```
Comment (codefromthecrypt) [inline internal/wasm/jit/arch_other.go:5]:
```suggestion
// archContext is empty on an unsupported architecture.
type archContext struct{}
```

File: internal/wasm/jit/engine.go [hunk 6]
```diff
@@ -810,13 +815,17 @@ func compileHostFunction(f *wasm.FunctionInstance) (*compiledFunction, error) {
 }
 
 func compileWasmFunction(f *wasm.FunctionInstance) (*compiledFunction, error) {
+	// golang-asm is not goroutine-safe so we take lock until we complete the compilation.
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:818]: move this comment to assemblerMutex instead of repeating.

**Review by codefromthecrypt — APPROVED** [review]
Great!

**Comment by mathetake** [comment]
windows failure seems persistent..

**Comment by mathetake** [comment]
Seems like this is due to the notorious `TestModuleEngine_Close_Concurrent`.. I will open an issue 

```
fatal error: unexpected signal during runtime execution
[signal 0xc0000005 code=0x1 addr=0x2cbf0fedd28 pc=0x8fc66a]

goroutine 22 [running]:
runtime.throw({0xbbcef2, 0x2cbf0fedd[18](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:18)})
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/panic.go:1[19](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:19)8 +0x76 fp=0xc0005397c0 sp=0xc000539790 pc=0x908eb6
runtime.sigpanic()
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/signal_windows.go:248 +0x22b fp=0xc000539808 sp=0xc0005397c0 pc=0x91d60b
runtime.newBucket(0x1, 0x9)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/mprof.go:175 +0x6a fp=0xc000539838 sp=0xc000539808 pc=0x8fc66a
runtime.stkbucket(0x1, 0xd0, {0xc0005398e8, 0xc0005396f0, 0x[20](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:20)}, 0x1)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/mprof.go:240 +0x1f4 fp=0xc0005398a0 sp=0xc000539838 pc=0x8fc974
runtime.mProf_Malloc(0xc000022340, 0xd0)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/mprof.go:344 +0xb1 fp=0xc000539a10 sp=0xc0005398a0 pc=0x8fce91
runtime.profilealloc(0x2cbcbac4b0d, 0xd0, 0x8e6e2f)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/malloc.go:1270 +0x85 fp=0xc000539a48 sp=0xc000539a10 pc=0x8ddb05
runtime.mallocgc(0xd0, 0xb80120, 0x1)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/malloc.go:1149 +0x72a fp=0xc000539ac8 sp=0xc000539a48 pc=0x8dd6aa
runtime.newobject(0x8dd634)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/malloc.go:1234 +0x27 fp=0xc000539af0 sp=0xc000539ac8 pc=0x8dd987
runtime.CallersFrames(...)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/symtab.go:66
runtime.Caller(0x2)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/extern.go:203 +0x7e fp=0xc000539c00 sp=0xc000539af0 pc=0x8d9ebe
github.com/stretchr/testify/assert.CallerInfo()
	C:/Users/runneradmin/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:115 +0x6c fp=0xc000539cc8 sp=0xc000539c00 pc=0xaf5fac
github.com/stretchr/testify/assert.Fail({0x2cbf0e8e2c8, 0xc00037e340}, {0xbb8ff4, 0x[21](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:21)}, {0x0, 0x0, 0x0})
	C:/Users/runneradmin/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:[24](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:24)6 +0x89 fp=0xc000539d98 sp=0xc000539cc8 pc=0xaf6929
github.com/stretchr/testify/assert.Error({0x2cbf0e8e2c8, 0xc00037e340}, {0x0, 0x0}, {0x0, 0x0, 0x0})
	C:/Users/runneradmin/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:1349 +0x9d fp=0xc000539de0 sp=0xc000539d98 pc=0xaf9d7d
github.com/stretchr/testify/require.Error({0xc19[25](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:25)8, 0xc00037e340}, {0x0, 0x0}, {0x0, 0x0, 0x0})
	C:/Users/runneradmin/go/pkg/mod/github.com/stretchr/testify@v1.7.0/require/require.go:253 +0x96 fp=0xc000539e[28](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:28) sp=0xc000539de0 pc=0xafb476
github.com/tetratelabs/wazero/internal/wasm/jit.TestModuleEngine_Close_Concurrent.func2(0xc00037e340)
	D:/a/wazero/wazero/internal/wasm/jit/engine_test.go:361 +0xaa7 fp=0xc000539f70 sp=0xc000539e28 pc=0xb15507
testing.tRunner(0xc00037e340, 0xc00004c780)
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/testing/testing.go:1259 +0x102 fp=0xc000539fc0 sp=0xc000539f70 pc=0x9a9202
testing.(*T).Run·dwrap·21()
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/testing/testing.go:1[30](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:30)6 +0x2a fp=0xc000539fe0 sp=0xc000539fc0 pc=0x9a9f0a
runtime.goexit()
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc000539fe8 sp=0xc000539fe0 pc=0x9396e1
created by testing.(*T).Run
	C:/hostedtoolcache/windows/go/1.17.7/x64/src/testing/testing.go:1306 +0x[35](https://github.com/tetratelabs/wazero/runs/5563644494?check_suite_focus=true#step:5:35)a
```

**Pull request merged**